lib/segment/src/entry/entry_point.rs

```diff
@@ -316,8 +316,74 @@ pub trait SegmentEntry {
         snapshotted_segments: &mut HashSet<String>,
     ) -> OperationResult<()>;
 
+    fn take_partial_snapshot(
+        &self,
+        temp_path: &Path,
+        tar: &tar_ext::BuilderExt,
+        manifest: &SegmentManifest,
+    ) -> OperationResult<()>;
+
+    fn get_segment_manifest(&self) -> OperationResult<SegmentManifest>;
+
     // Get collected telemetry data of segment
     fn get_telemetry_data(&self, detail: TelemetryDetail) -> SegmentTelemetry;
 
     fn fill_query_context(&self, query_context: &mut QueryContext);
 }
+
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct SegmentManifest {
+    pub segment_version: SeqNumberType,
+    pub file_versions: HashMap<PathBuf, FileVersion>,
+}
+
+#[derive(
+    Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, serde::Serialize, serde::Deserialize,
+)]
+#[serde(untagged)]
+pub enum FileVersion {
+    Version(SeqNumberType),
+    Unversioned,
+}
+
+impl FileVersion {
+    pub fn is_unversioned(self) -> bool {
+        self == Self::Unversioned
+    }
+
+    pub fn or_segment_version(self, segment_version: SeqNumberType) -> SeqNumberType {
+        match self {
+            FileVersion::Version(version) => version,
+            FileVersion::Unversioned => segment_version,
+        }
+    }
+}
+
+impl From<SeqNumberType> for FileVersion {
+    fn from(version: SeqNumberType) -> Self {
+        Self::Version(version)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn file_version_serde() {
+        test_file_version_serde(FileVersion::Version(42), "42");
+        test_file_version_serde(FileVersion::Unversioned, "null");
+    }
+
+    fn test_file_version_serde(version: FileVersion, json: &str) {
```

Comment:
I was a tiny bit confused on what this function did, may be a comment helps:

```suggestion
    /// Test that file version variants are uniquely represented in JSON
    fn test_file_version_serde(version: FileVersion, json: &str) {
```

Reply:
Done.